pandas/core/indexes/base.py

```diff
@@ -6041,12 +6041,13 @@ def get_indexer_non_unique(
 
         # Note: _maybe_downcast_for_indexing ensures we never get here
         #  with MultiIndex self and non-Multi target
-        tgt_values = target._get_engine_target()
         if self._is_multi and target._is_multi:
             engine = self._engine
             # Item "IndexEngine" of "Union[IndexEngine, ExtensionEngine]" has
             # no attribute "_extract_level_codes"
             tgt_values = engine._extract_level_codes(target)  # type: ignore[union-attr]
+        else:
+            tgt_values = target._get_engine_target()
```

Comment:
I think its fine to move this line into the `else` block as you've done, but I'm not seeing any impact on performance:

```
import pandas as pd
import numpy as np

N = 1_000_000

mi1 = pd.MultiIndex.from_arrays([np.ones(N), np.ones(N)])
mi2 = pd.MultiIndex.from_tuples([(1, 1)])

%timeit mi1.get_indexer_non_unique(mi2)

# 953 ms ± 29.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  -> main
# 949 ms ± 29.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  -> PR
```

Do you have an example where this impacts performance?